micropip/package_index.py

```diff
@@ -0,0 +1,188 @@
+
+    @staticmethod
+    def from_simple_api(data: dict[str, Any]) -> "ProjectInfo":
+        """
+        Parse Simple API response
+
+        https://peps.python.org/pep-0503/
+        https://peps.python.org/pep-0691/
+        """
+        name = data["name"]
+
+        # List of versions (PEP 700), this key is not critical to find packages
+        # but it is required to ensure that the same class instance is returned
+        # from JSON and Simple APIs.
+        versions = data.get("versions", [])
+
+        # Group files by version
+        releases: dict[Version, list[Any]] = defaultdict(list)
+
+        for version in versions:
+            if not _is_valid_pep440_version(version):
+                continue
+
+            releases[Version(version)] = []
+
+        for file in data["files"]:
+            filename = file["filename"]
+
+            if not _fast_check_incompatibility(filename):
+                # parsing a wheel filename is expensive, so we do a quick check first
+                continue
+
+            try:
+                version = parse_version(filename)
+            except (InvalidVersion, InvalidWheelFilename):
+                continue
+
+            releases[version].append(file)
+
+        return ProjectInfo._compatible_only(name, releases)
+
+    @staticmethod
+    def _compatible_only(
+        name: str, releases: dict[Version, list[dict[str, Any]]]
+    ) -> "ProjectInfo":
+        def _compatible_wheels(
+            files: list[dict[str, Any]], version: Version
+        ) -> Generator[ProjectInfoFile, None, None]:
+            """
+            Return a generator of wheels compatible with the current platform.
+            Checking compatibility takes a bit of time, so we use a generator to avoid doing it if not needed.
+            """
+            for file in files:
+                filename = file["filename"]
+
+                # Checking compatibility takes a bit of time,
+                # so we use a generator to avoid doing it for all files.
+                compatible = is_package_compatible(filename)
+                if not compatible:
+                    continue
+
+                # JSON API has a "digests" key, while Simple API has a "hashes" key.
+                hashes = file["digests"] if "digests" in file else file["hashes"]
+
+                # TODO: For now we expect that the sha256 hash is always available.
+                # This is true for PyPI, but may not be true for other package indexes,
+                # since it is not a hard requirement of PEP503.
+                sha256 = hashes["sha256"]
+
+                yield ProjectInfoFile(
+                    filename=filename,
+                    url=file["url"],
+                    version=version,
+                    sha256=sha256,
+                    size=file.get("size"),
+                )
+
+        releases_compatible = {
+            version: _compatible_wheels(files, version)
```

Comment:
If we want to go faster (no need to do it now). We can probably also subclass Version to something that has a frozen __str__ and __hash__ so it doesn't need to be re-computed at each dict comparison..

Reply:
I opened an issue (https://github.com/pyodide/micropip/issues/73) so that we can do it as a follow-up. I also think it is would be a good issue for new contributors if there is other sprint / tutorial events.